src/com/dosse/upnp/Gateway.java

```diff
@@ -117,12 +114,12 @@ public Gateway(byte[] data, Inet4Address ip) throws Exception {
         conn.setRequestProperty("Content-Length", "" + req.length);
         conn.getOutputStream().write(req);
         Document d = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(conn.getInputStream());
-        NodeIterator iter = ((DocumentTraversal) d).createNodeIterator(d.getDocumentElement(), NodeFilter.SHOW_ELEMENT, null, true);
-        Node n;
-        while ((n = iter.nextNode()) != null) {
+        NodeList nodeList = d.getElementsByTagName("*");
+        for (int i = 0; i < nodeList.getLength(); i++) {
             try {
-                if (n.getFirstChild().getNodeType() == Node.TEXT_NODE) {
-                    ret.put(n.getNodeName(), n.getTextContent());
+                Node node = nodeList.item(i);
+                if (node.getNodeType() == Node.ELEMENT_NODE) {
```

Comment:
Shouldn't this be Node.TEXT_NODE?

Reply:
No, getElementsByTagName returns element nodes.

BTW I found this solution here: https://stackoverflow.com/questions/5386991/java-most-efficient-method-to-iterate-over-all-elements-in-a-org-w3c-dom-docume#answer-5511298